python/cugraph-equivariant/cugraph_equivariant/nn/tensor_product_conv.py

```diff
@@ -0,0 +1,252 @@
+        edge_sh: torch.Tensor,
+        edge_emb: torch.Tensor,
+        graph: tuple[torch.Tensor, tuple[int, int]],  # COO, (n_src, n_dst)
+        src_scalars: Optional[torch.Tensor] = None,
+        dst_scalars: Optional[torch.Tensor] = None,
+        reduce: str = "mean",
+        edge_envelope: Optional[torch.Tensor] = None,
+    ) -> torch.Tensor:
+        """Forward pass.
+
+        Parameters
+        ----------
+        src_features : torch.Tensor
+            Source node features.
+            Shape: (num_src_nodes, in_irreps.dim)
+
+        edge_sh : torch.Tensor
+            The spherical harmonic representations of the edge vectors.
+            Shape: (num_edges, sh_irreps.dim)
+
+        edge_emb: torch.Tensor
+            Edge embeddings that are fed into MLPs to generate tensor product weights.
+            Shape: (num_edges, dim), where `dim` should equal to:
+            - `tp.weight_numel` when the layer does not contain MLPs.
+            - `num_scalars` when `mlp_fast_first_layer` enabled.
+            - `mlp_channels[0]` otherwise.
+
+        graph : Any
+            The graph.
+
+        src_scalars: torch.Tensor, optional
+            Scalar features of source nodes.
+            Shape: (num_src_nodes, num_scalars)
+
+        dst_scalars: torch.Tensor, optional
+            Scalar features of destination nodes.
+            Shape: (num_dst_nodes, num_scalars)
+
+        reduce : str, optional (default="mean")
+            Reduction operator. Choose between "mean" and "sum".
+
+        edge_envelope: torch.Tensor, optional
+            Edge envelope.
+            Shape: (num_edges,)
+
+        Returns
+        -------
+        torch.Tensor
+            Output node features.
+            Shape: (num_dst_nodes, out_irreps.dim)
+        """
+        if self.mlp is None:
+            assert self.tp.weight_numel == edge_emb.size(-1)
+        else:
+            if self.mlp_fast_first_layer:
+                assert edge_emb.size(-1) == self.num_scalars
+                assert src_scalars.size(-1) == self.num_scalars
+                assert dst_scalars.size(-1) == self.num_scalars
+            else:
+                assert self.mlp[0].in_features == edge_emb.size(-1)
+
+        if reduce not in ["mean", "sum"]:
+            raise ValueError(
+                f"reduce argument must be either 'mean' or 'sum', got {reduce}."
+            )
+
+        (src, dst), (num_src_nodes, num_dst_nodes) = graph
+
+        if self.mlp is not None:
+            if self.mlp_fast_first_layer:
+                w_edge, w_src, w_dst = torch.chunk(self.mlp[0].weight, chunks=3, dim=-1)
+                tp_weights = (
+                    edge_emb @ w_edge.T
+                    + (src_scalars @ w_src.T)[src]
+                    + (dst_scalars @ w_dst.T)[dst]
+                    + self.mlp[0].bias
+                )
+                tp_weights = self.mlp[1:](tp_weights)
+            else:
+                tp_weights = self.mlp(edge_emb)
```

Comment:
Here I would either (i) raise an error if `src_scalars` or `dst_scalars` are provided but `self.mlp_fast_first_layer` is False or (ii) concatenate `src_scalars[src]` and `dst_scalars[dst]` into `edge_emb` in the `else:` block

Reply:
Will remove `mlp_fast_first_layer` option here, and trigger the faster GEMM based on inputs from forward.